# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: extensions/permissions/PermissionDelegateHandler.cpp
# Commit: 4e6d8f786f87
# Full Hash: 4e6d8f786f87cca483cd5f54f9f57d2519eceb16
# Author: Tim Huang <tihuang@mozilla.com>
# Date: 2020-06-13 03:21:52
# Regressor Bug: 1587743
# File Overlap Count: 1
# Description:
#   Bug 1587743 - Part 3: Using WindowContext to test delegated permissions. r=baku
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D79134
# ==============================================================================

diff -r a2bec965260d -r 4e6d8f786f87 extensions/permissions/PermissionDelegateHandler.cpp
--- a/extensions/permissions/PermissionDelegateHandler.cpp	Fri Jun 12 16:31:56 2020 +0000
+++ b/extensions/permissions/PermissionDelegateHandler.cpp	Fri Jun 12 16:32:03 2020 +0000
@@ -216,6 +216,7 @@
                                                   uint32_t* aPermission,
                                                   bool aExactHostMatch) {
   MOZ_ASSERT(mDocument);
+  MOZ_ASSERT(mPrincipal);
 
   if (mPrincipal->IsSystemPrincipal()) {
     *aPermission = nsIPermissionManager::ALLOW_ACTION;
@@ -250,7 +251,32 @@
   if ((info->mPolicy == DelegatePolicy::eDelegateUseTopOrigin ||
        (info->mPolicy == DelegatePolicy::eDelegateUseFeaturePolicy &&
         StaticPrefs::dom_security_featurePolicy_enabled()))) {
-    // TODO: This will be changed in the next patch.
+    RefPtr<WindowContext> topWC =
+        mDocument->GetWindowContext()->TopWindowContext();
+
+    if (topWC->IsInProcess()) {
+      // If the top-level window context is in the same process, we directly get
+      // the node principal from the top-level document to test the permission.
+      // We cannot check the lists in the window context in this case since the
+      // 'perm-changed' could be notified in the iframe before the top-level in
+      // certain cases, for example, request permissions in first-party iframes.
+      // In this case, the list in window context hasn't gotten updated, so it
+      // would has an out-dated value until the top-level window get the
+      // observer. So, we have to test permission manager directly if we can.
+      RefPtr<Document> topDoc = topWC->GetBrowsingContext()->GetDocument();
+
+      if (topDoc) {
+        principal = topDoc->NodePrincipal();
+      }
+    } else {
+      // Get the delegated permissions from the top-level window context.
+      DelegatedPermissionList list =
+          aExactHostMatch ? topWC->GetDelegatedExactHostMatchPermissions()
+                          : topWC->GetDelegatedPermissions();
+      size_t idx = std::distance(sPermissionsMap, info);
+      *aPermission = list.mPermissions[idx];
+      return NS_OK;
+    }
   }
 
   return (mPermissionManager->*testPermission)(principal, aType, aPermission);
